src/lib/structures/modelMap.ml

```diff
@@ -30,100 +30,182 @@
-
 type t = {
-  values : V.t P.t;
+  values : Graph.t P.t;
   suspicious : bool;
 }
 
-let is_suspicious_name hs =
-  match Hstring.view hs with
-  | "@/" | "@%" | "@*" -> true
-  | _ -> false
-
-(* The model generation is known to be imcomplete for FPA and Bitvector
-   theories. *)
-let is_suspicious_symbol = function
-  | Sy.Op (Float | Fixed | Abs_int | Abs_real | Sqrt_real
-          | Sqrt_real_default | Sqrt_real_excess
-          | Real_of_int | Int_floor | Int_ceil
-          | Max_int | Max_real | Min_int | Min_real
-          | Pow | Integer_log2 | Int2BV _ | BV2Nat
-          | BVand | BVor | Integer_round) -> true
-  | Sy.Name (hs, _, _) when is_suspicious_name hs -> true
-  | _ -> false
-
-let add ((sy, _, _) as p) v { values; suspicious } =
-  let prof_p = try P.find p values with Not_found -> V.empty in
-  let values =
-    if V.mem v prof_p then values
-    else P.add p (V.add v prof_p) values
-  in
-  { values; suspicious = suspicious || is_suspicious_symbol sy }
-
-let iter f { values; _ } =
-  P.iter (fun ((sy, _, _) as key) value ->
-      match sy with
-      | Sy.Name (_, _, true) ->
-        (* We don't print constants defined by the user. *)
-        ()
-      | _ -> f key value
-    ) values
-
-let[@inline always] is_suspicious { suspicious; _ } = suspicious
-
-let fold f { values; _ } acc = P.fold f values acc
-
-let empty = { values = P.empty; suspicious = false }
-
-let is_empty { values; _ } = P.is_empty values
+let add ((id, arg_tys, _) as sy) arg_vals ret_val { values; suspicious } =
+  if List.compare_lengths arg_tys arg_vals <> 0 then
+    Fmt.invalid_arg "The arity of the symbol %a doesn't agree the number of \
+                     arguments" Id.pp id;
+
+  let graph = try P.find sy values with Not_found -> Graph.empty in
+  let values = P.add sy (Graph.add arg_vals ret_val graph) values in
+  { values; suspicious }
+
+let empty ~suspicious = { values = P.empty; suspicious }
+
+let pp_named_arg_ty ~unused ppf (arg_name, arg_ty) =
+  let pp_unused ppf unused = if unused then Fmt.pf ppf "_" else () in
+  Fmt.pf ppf "(%aarg_%i %a)" pp_unused unused arg_name Ty.pp_smtlib arg_ty
+
+let pp_define_fun ppf ((id, arg_tys, ret_ty), graph) =
+  let inverse_rel = Graph.inverse graph in
+  let is_constant = Value.Map.cardinal inverse_rel = 1 in
+  let named_arg_tys = List.mapi (fun i arg_ty -> (i, arg_ty)) arg_tys in
+  Fmt.pf ppf "(@[define-fun %a (%a) %a@ %a)@]"
+    Id.pp id
+    Fmt.(list ~sep:sp (pp_named_arg_ty ~unused:is_constant)) named_arg_tys
+    Ty.pp_smtlib ret_ty
+    Graph.pp_inverse inverse_rel
+
+let pp_seq ppf seq =
+  match seq () with
+  | Seq.Nil -> ()
+  | Cons _ -> Fmt.pf ppf "@,%a" Fmt.(seq ~sep:cut pp_define_fun) seq
+
+let pp ppf {values; suspicious} =
+  if suspicious then begin
+    Fmt.pf ppf "; This model is a best-effort. It includes symbols\n\
+                ; for which model generation is known to be incomplete.@."
+  end;
+  Fmt.pf ppf "@[<v 2>(%a@]@,)" pp_seq (P.to_seq values)
```

Comment:
We can check `P.is_empty` and then `Fmt.iter P.iter` rather than converting to a sequence.